programs/benchmark/Benchmark.cpp

```diff
@@ -533,29 +613,82 @@ class Benchmark : public Poco::Util::Application
 
         executor.finish();
 
+        watch.stop();
         double duration = (display_client_side_time || progress.elapsed_ns == 0)
             ? watch.elapsedSeconds()
             : progress.elapsed_ns / 1e9;
+        size_t info_index = round_robin ? 0 : connection_index;
 
-        std::lock_guard lock(mutex);
+        if (precise && cur_interval)
+        {
+            // Stats weighting across all overlapped intervals
+            UInt64 duration_ns = watch.getEnd() - watch.getStart();
+
+            std::lock_guard lock(interval_mutex);
+
+            // Distribute weights across intervals intersecting query execution span
+            // Intervals:
+            // [beg]    I0: [s0---------e0)
+            //          I1:               [s1--------------e1)
+            // [end]    I2:               |                  [s2-----------e2)
+            //                            |                  |
+            //  Query span:        [qs----+------------------+-------qe)
+            //                         ^           ^              ^
+            //                         w0          w1             w2
+            // Weights (w0, w1, w2) are proportional to the lengths of intersections
+            // and normalized to the total duration of a query to given 1 as a sum of weights.
+            for (; cur_interval; cur_interval = cur_interval->next)
+            {
+                if (cur_interval->end_ns <= watch.getStart())
+                    continue;
+                if (cur_interval->start_ns >= watch.getEnd())
+                    break;
+                double overlap_ns = std::min(cur_interval->end_ns, watch.getEnd()) - std::max(cur_interval->start_ns, watch.getStart());
+                double weight = double(overlap_ns) / duration_ns;
```

Comment:
`nit:` casting to double is superflous here since overlap_ns is already double.
```suggestion
                double weight = overlap_ns / duration_ns;
```

Reply:
Ah, yes, thank you